### ISSUE TITLE
bug(elements): importing `createCustomElement` breaks SSR

### ISSUE THREAD

**Issue by thefliik** [angular/angular#24551]
## I'm submitting a...

<pre><code>
[x] Bug report  
</code></pre>

## Current behavior

Importing `createCustomElement` throws an error during angular universal server side rendering.

Note: it appears that simply importing the function throws the error. Currently, even the following code throws an error during SSR:

```
import {createCustomElement} from '@angular/elements';
. . .
export class AppModule {
  constructor() {
    if (false as any) {
      createCustomElement
    }
  }
}
```
#### Error:
```
var elProto = Element.prototype;
              ^

ReferenceError: Element is not defined
    at Object.<anonymous> (/user/apps/tests/universal-starter/dist/server.js:126911:15)
    at __webpack_require__ (/user/apps/tests/universal-starter/dist/server.js:20:30)
    at Object.@angular/elements (/user/apps/tests/universal-starter/dist/server.js:126796:18)
    at __webpack_require__ (/user/apps/tests/universal-starter/dist/server.js:126326:30)
    at Object../src/app/app.module.ts (/user/apps/tests/universal-starter/dist/server.js:126453:18)
    at __webpack_require__ (/user/apps/tests/universal-starter/dist/server.js:126326:30)
    at Object../src/app/app.server.module.ngfactory.js (/user/apps/tests/universal-starter/dist/server.js:126504:11)
    at __webpack_require__ (/user/apps/tests/universal-starter/dist/server.js:126326:30)
    at Object../src/main.server.ts (/user/apps/tests/universal-starter/dist/server.js:126712:37)
    at __webpack_require__ (/user/apps/tests/universal-starter/dist/server.js:126326:30)
```

## Expected behavior

Ideally, custom elements could be rendered via SSR normally, but, at minimum, it'd expect the ability to import `createCustomElement` and then *conditionally call it* using `isPlatformBrowser()` so that the custom element was only defined + registered on the client side.

## Minimal reproduction of the problem with instructions

A reproduction of the issue can be found here: https://github.com/thefliik/universal-starter/tree/elements-ssr-bug. Simply clone the repo and run `yarn build:ssr && yarn serve:ssr`.

The repo was made by cloning the `@angular/universal-starter` repo, adding `@angular/elements` to `project.json`, and updating `app.module.ts` with:

```
import {createCustomElement} from '@angular/elements';

...

export class AppModule {
  constructor() {
    /**
     * # Note
     * 
     * Remove this statement and no errors are thrown. I *think* this is because
     * tree shaking will remove the `createCustomElement` import statement without
     * this line.
     */
    if (false as any) {
      createCustomElement
    }
  }
}
```

## What is the motivation / use case for changing the behavior?

To allow custom elements built with `@angular/elements` to be used in a SSR app.

My specific use case for angular elements is as a work around for issue #14324 (& #17168).

## Environment

<pre><code>
Angular CLI: 6.0.0
Node: 8.9.1
OS: darwin x64
Angular: 6.0.5
... animations, common, compiler, core, elements, forms, http
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.6.0
@angular-devkit/build-angular     0.6.0
@angular-devkit/build-optimizer   0.6.0
@angular-devkit/core              0.6.0
@angular-devkit/schematics        0.6.0
@angular/cli                      6.0.0
@angular/compiler-cli             6.0.0
@angular/language-service         6.0.0
@ngtools/webpack                  6.0.0
@schematics/angular               0.6.0
@schematics/update                0.6.0
rxjs                              6.0.0
typescript                        2.7.2
webpack                           4.6.0
</code></pre>

**Comment by thefliik** [comment]
As a work around, you can conditionally import `createCustomElement` with `const { createCustomElement } = require('@angular/elements')`.

For example:

```
  constructor(private injector: Injector, @Inject(PLATFORM_ID) platformId: Object) {
    if (isPlatformBrowser(platformId)) {
      const { createCustomElement } = require('@angular/elements');

      // register the custom element with the browser.
      customElements.define(
        'app-calendar-breadcrumb',
        createCustomElement(CalendarBreadcrumbComponent, {injector: this.injector})
      );
    }
  }
```

**Comment by ericmartinezr** [comment]
Try the suggestion in this issue https://github.com/angular/angular/issues/24345

**Comment by thefliik** [comment]
@ericmartinezr thanks for the suggestion. I had tried the solution suggested in that issue before reporting this and, while it does eliminate the `var elProto = Element.prototype;` error, it results in a new error. I don't have a copy of the error in front of me, but it was something along the lines of: _~"readonly property `constructor` cannot be reassigned"_

**Comment by tnicola** [comment]
Same issue here, using @thefliik workaround it works:
```
if (isPlatformBrowser(platformId)) {
      const { createCustomElement } = require('@angular/elements');
      // using createCustomElement 
}
```

**Comment by trotyl** [comment]
It's actually a problem of Angular CLI, for server-side rendering for custom elements, it should always use `"target": "es2015"` rather than requiring the polyfills (which is not compatible with node).

But currently Angular CLI will always skip bundle of `@angular` packages and forwarding to UMD bundles, which are es5 code which doesn't fulfill Custom Elements requirement (real `class` type).

**Comment by thefliik** [comment]
Should I re-report this issue in the angular-cli repo?

**Comment by trotyl** [comment]
I think it's also a question about the workflow:

**Should Angular Elements be rendered on the server side?**

An Angular Element is just a normal Custom Element, which is bootstrapped by browser, not Angular.

Say that one have:

```typescript
@Component({
  selector: 'my-comp',
  template: `
    <p>Name: {{ name }}</p>
    <slot></slot>
  `
})
class MyComp {}
```

Used by:

```html
<my-comp name="Foo">Bar</my-comp>
```

Got rendered as static HTML on the server which will be consumed by browser:

```html
<my-comp ng-version="6.0.0" name="Foo">
  <p>Foo</p>
  Bar
</my-comp>
```

And then **in the browser, the previous render result on the server will be treated as content**, results to:

```html
<my-comp ng-version="6.0.0" name="Foo">
  <p>Foo</p>
  <p>Foo</p>
  Bar
</my-comp>
```

Which is certainly unexpected result.

My opinion is:

**Without a perfect re-hydration process (which doesn't exist in Angular yet), the server-side rendering should never take Angular Elements into consideration, but just left to browser.**

**Comment by kevinphelps** [comment]
@trotyl: That scenario can be solved by not using the same selector for the Angular component and the custom element. This seems to work well for using custom elements in a CMS-type system. The CMS content simply uses a different prefix than the component selectors.

**Comment by trotyl** [comment]
@kevinphelps It's unrelated to selector, as long as custom elements being recognized in domino, then it triggers Angular Elements and got broken. There's no way to render `<slot>` or `<style>` in a shadow DOM to static HTML. (Of course there's no issue if you don't use shadow DOM)

**Comment by kevinphelps** [comment]
I was referring to the content nesting.

**Comment by kevinphelps** [comment]
I'm trying to find a way to render custom elements in node without resorting to using puppeteer. Haven't figured it out yet, lol.

**Comment by trotyl** [comment]
@kevinphelps I guess you are using Angular Elements inside an Angular application, that does "work", not due to Angular Elements can be properly handled, but the fact an Angular application would clear all existing HTML content in the bootstrapped element, and custom elements registration happens during the bootstrap, so browser never had a chance to recognize server-rendered custom elements.

**Comment by kevinphelps** [comment]
How do you server render custom elements? I haven't figured that out yet, but I see lots of comments around GitHub implying that it's possible.

**Comment by trotyl** [comment]
@kevinphelps AFAIK the best approach is something like https://github.com/skatejs/skatejs/tree/master/packages/ssr, which adding additional `<shadow-root>` placeholder elements and inline scripts to covert them to real shadowRoots.

So that it could work with hydration with all the DOM nodes reused between server and client.

**Comment by gkalpak** [comment]
FWIW, this is still an issue with v9 (and v10-rc.1).

For future reference, I tried to change [this line][1] to `global.Element.prototype` (`global` being [imported from `core`][2], but the error remained. I was hoping that [this line][3] would take care of exposing `domino.impl.Element` as `global.Element`, but that function is probably not called in time to avoid the error.

More investigation is needed on determining the exact root cause.

[1]: https://github.com/angular/angular/blob/78ea2527a197f75312b4092083426c342eee5e13/packages/elements/src/utils.ts#L11
[2]: https://github.com/angular/angular/blob/78ea2527a197f75312b4092083426c342eee5e13/packages/core/src/util/global.ts#L27
[3]: https://github.com/angular/angular/blob/78ea2527a197f75312b4092083426c342eee5e13/packages/platform-server/src/domino_adapter.ts#L19

**Comment by vinothbabu** [comment]
Is there any hack or solution to support custom-elements in universal?

**Comment by gkalpak** [comment]
To solve this particular issue, we need a fix along the lines of #37799.
